### ISSUE TITLE
coeff used by (correctly) not implemented for spoly

### ISSUE THREAD

**Issue by wbhart** [oscar-system/Singular.jl#398]
spoly cannot provide a coeff(f, n) function due to the fact that Singular does not have random access structures for polynomials.

This means the following are implemented incorrectly in Singular (or no longer needed):

canonical_unit(a::spoly{T})  -- src/poly/poly.jl
isnegative(x::spoly) -- src/poly/poly.jl

**Comment by wbhart** [comment]
In both cases coeff(f, 0) can be replaced with first(coefficients(f)), which will give the leading coefficient, which is presumably what was meant. Alternatively, perhaps leading_coefficient is better once the relevant PR is merged.

**Comment by fieker** [comment]
There is a chance that in Hecke at some point we (I) added hasLength to the iterator

**Comment by wbhart** [comment]
I am a bit slow. I finally figured out how this is suddenly a problem now. The coefficients iterator in Hecke was always called coefficients. The one in AbstractAlgebra was called coeffs. Now that the latter has been renamed to coefficients it is overwriting the AbstractAlgebra version.

So the solution is to remove the iterator from Hecke, which is no longer needed, as it does precisely what the AbstractAlgebra one does, length and all. Singular has its own iterators which override the AbstractAlgebra ones and do not rely on length.

I will make a PR to remove this from Hecke after testing that it still passes its tests without it.

**Comment by thofma** [comment]
I am already on it.

**Comment by thofma** [comment]
I will ping you.

**Comment by wbhart** [comment]
Ah ok. Also note the lead function immediately before it is no longer needed (presumably). I am not sure how I missed this one.

**Comment by thofma** [comment]
Yeah I saw it. I will remove some stuff and double check with you.

**Comment by wbhart** [comment]
Oh, I am posting on the wrong issue anyway, sorry.

**Comment by wbhart** [comment]
Just to clarify, addition of has_length wouldn't make these definitions correct, even if Singular depended on Hecke. Fortunately with the new printing code, these may not be needed any more anyway.
